Fixed race condition in ProcessManager

dmatej opened the pull request:
- reader.readLine can return null when there was nothing to read.
- escallate errors to logs, don't hide them.
- fixes issue discovered in https://github.com/eclipse-ee4j/glassfish/pull/24446#issuecomment-1585533239

dmatej committed:
Fixed race condition in ProcessManager

- reader.readLine can return null when there was nothing to read.
- escallate errors to logs, don't hide them.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

avpinchuk commented on nucleus/common/common-util/src/main/java/com/sun/enterprise/universal/process/ProcessManager.java:
```diff
@@ -222,51 +211,66 @@ private void waitAwhile(Process process) throws InterruptedException, ProcessMan
-        private final StringBuffer sb;
+        private final StringBuilder sb;
         private final boolean echo;
+        private boolean stop;
```
Hmm, non-volatile flag enough here?

dmatej replied:
I believe it is not required here. If it would be several nanoseconds later it doesn't matter. Perhaps we should make some experiments if it is better to use volatile and slow down some context switching or not to use volatile and slow down the "shutdown" (it can make few iterations more).

avpinchuk replied:
The `stop` flag set to true in the thread running `execute` (in `finish` method), but check in another (reader) thread. I think, visibility problem is possible here. We can get an infinite wait loop.

dmatej replied:
Volatile is not a solution for thread safety, just an instruction to avoid optimization. The value change has just one direction, from false to true, and then with volatile it must be checked in memory before every use, without volatile it can be used from cpu cache between syncing the value state in memory, which is done by the cpu.
Or am I wrong and cpu can use the old value forever?
Can you prove that behavior on some cpu?

avpinchuk replied:
Volatile write give us StoreLoad barrier to flush CPU write buffers. Without it cpu may use value from its local wtite buffer potentially indefinitely, i.e., second thread, running on another core/cpu may never see new value.

I have no cpu other than x86 (amd and intel) ;-(

dmatej replied:
Hmm, I know about related issues if we would have two depending values, but in this case it seems contraproductive to me. But we don't need to hunt nanos here, so for simplicity I can just add it.

I am just curious if we are not overengineering that. Just to remember ie if arms would have such issue... and how to reproduce it.

avpinchuk replied:
In case of two variables even volatile may not be sufficient ;-)

avpinchuk replied:
As a microoptimization we can use AtomicBoolen lazySet/get here

dmatej replied:
I had to read more ... now I am thinking about some tests outside GF related to this problematic :-)
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/invoke/VarHandle.html
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/atomic/AtomicBoolean.html

I used acquire-release, finally.
I tried to google where we could reproduce some issues like never ending loop, somebody mentioned Apple and ARM cpus.

avpinchuk replied:
Great! I think, release/acquire enough here.

dmatej committed:
Unused waitForReaderThreads removed

- it was an old hack for windows to avoid an old fixed bug (issue numbers lost)

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Using AtomicBoolean instead of nonvolatile boolean

- Should be more reliable on Mac and ARM cpus (not tested yet)

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed mistake from recent commit - removal of waitForReaderThreads

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

avpinchuk commented:
I clone your branch and still get same error https://github.com/eclipse-ee4j/glassfish/pull/24446#issuecomment-1585533239

dmatej committed:
Fixed processing of process output

- The output might be unavailable after the forcible process termination
- await methods ignored return values
- Shortened timeout for log processing to 1 second.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Yet more fixes and optimizations

- log formatting (makes the test nearly twice faster on my machine)
- don't destroy process if it already exited
- testSetStdinLines runs 1000 times now, because the error was not much
  reproducible (0.1% after last commit)

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej commented:
Thanks for the navigation, @avpinchuk! What we discovered yet in private discussion - the manager in some parts ignored returned values, in another tried to kill dead process and THEN process the output, in another - for some reason the input values were not processed, but the error was reproducible just in some 0.1% on my machine.
I hope we resolved all of those issues now ...

avpinchuk commented:
@dmatej, happy to help ;-)

dmatej committed:
Removed final (injected value)

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>
